src/constr_gen.erl

```diff
@@ -324,32 +331,86 @@ bound_vars_pat(P) ->
     end.
 
 
+% ty_of_pat
 % \lbag p \rbag_\Gamma
--spec ty_of_pat(constr:constr_env(), ast:pat()) -> ast:ty().
-ty_of_pat(Env, P) ->
+%
+% In the paper, the type t of a pattern p has the following semantics:
+%     Expression e matches p if, and only if, e has type t
+%
+% With list patterns, this is no longer true.
+%
+% Example 1: pattern [1 | _].
+% For the => direction above, consider an expression e that matches
+% this pattern. From this, all we know is that e must have type nonempty_list(any()).
+% (e could be any of the following expressions: [1,2,3] or [1, "foo"] or [1]).
+% For the <= direction, e must have type nonempty_list(1) if we want to make sure
+% that the pattern definitely matches.
+% Example 2: pattern [_ | _ | _].
+% For the => direction, consider an expression e that matches the pattern.
+% From this, all we know is that e has type nonempy_list() because there is no
+% type for lists with at least length two.
+% For the <= direction, no type except none() guarantees that e matches the pattern.
+%
+% Hence, we introduce a mode for ty_of_pat, which can be either upper or lower.
+%
+% - Mode upper deals with the potential type. Any expression matching p must
+%   be of this type.
+%   Example 1: the potential type is nonempty_list(any()).
+%   Example 2: the potential type is nonempty_list(any()).
+%
+% - Mode lower deals with the accepting type. If e has this type, then p definitely
+%   matches.
+%   Example 1: the accepting type is nonempty_list(1).
+%   Example 2: the accepting type is none()
+-spec ty_of_pat(constr:constr_env(), ast:pat(), upper | lower) -> ast:ty().
+ty_of_pat(Env, P, Mode) ->
     case P of
         {'atom', _L, A} -> {singleton, A};
         {'char', _L, C} -> {singleton, C};
         {'integer', _L, I} -> {singleton, I};
         {'float', _L, _F} -> {predef, float};
         {'string', _L, _S} -> {predef_alias, string};
         {bin, L, _Elems} -> errors:unsupported(L, "bitstring patterns");
-        {match, _L, P1, P2} -> ast_lib:mk_intersection([ty_of_pat(Env, P1), ty_of_pat(Env, P2)]);
+        {match, _L, P1, P2} ->
+            ast_lib:mk_intersection([ty_of_pat(Env, P1, Mode), ty_of_pat(Env, P2, Mode)]);
         {nil, _L} -> {empty_list};
         {cons, _L, P1, P2} ->
-            %% FIXME: this is wrong. It should be the union of {list, ty_of_pat(Env, P1)}
-            %% and ty_of_pat(Env, P2), intersected with type list(any()).
-            %% Accepting and potential types of list patterns are also wrong.
-            ast_lib:mk_intersection([{list, ty_of_pat(Env, P1)}, ty_of_pat(Env, P2)]);
+            Symtab = symtab:empty(),  % FIXME: need to use a real symtab below?
```

Comment:
In the `typing:infer/2` function, there is a context `ctx` which includes the current Symtab. The call to `constr_gen:gen_constrs_fun_group/1` loses that symtab. The `ctx` type in `constr_gen` is missing the symtab. It should be added, and that symtab can (should) be used here.

Reply:
fixed